Userland/Libraries/LibJS/Runtime/VM.cpp

```diff
@@ -371,4 +381,44 @@ bool VM::in_strict_mode() const
     return call_frame().is_strict_mode;
 }
 
+void VM::run_queued_promise_jobs()
+{
+    dbgln_if(PROMISE_DEBUG, "Running queued promise jobs");
+    // Temporarily get rid of the exception, if any - job functions must be called
+    // either way, and that can't happen if we already have an exception stored.
+    TemporaryChange change(m_exception, static_cast<Exception*>(nullptr));
+    while (!m_promise_jobs.is_empty()) {
+        auto* job = m_promise_jobs.take_first();
+        dbgln_if(PROMISE_DEBUG, "Calling promise job function @ {}", job);
+        [[maybe_unused]] auto result = call(*job, js_undefined());
+    }
+    // Ensure no job has created a new exception, they must clean up after themselves.
+    VERIFY(!m_exception);
+}
+
+// 9.4.4 HostEnqueuePromiseJob, https://tc39.es/ecma262/#sec-hostenqueuepromisejob
+void VM::enqueue_promise_job(NativeFunction& job)
```

Comment:
I don't know if this should be done now, but this needs to somehow be overridable in the future for integration with HTML event loops:
https://html.spec.whatwg.org/#hostenqueuepromisejob

Reply:
Definitely one for later - from what I've seen LibWeb isn't driven by an event loop at all (yet), e.g. timers just run their callbacks right away instead of queuing a task, so I just embraced that and did this: 192f3a344477d6f09e42c533bbd7a92d50f30fa8

Since nothing would want to subclass `JS::VM` (probably), we might just add a `on_something` function and call that if it exists, leaving queueing and therefore running the job to whatever owns the VM.